### PR TITLE
fix(ai): bound local server responses

### PR DESCRIPTION
**Pull request by abechan1**
### Motivation

- Prevent unbounded memory and CPU use when querying a misbehaving local OpenAI-compatible server by bounding sizes for model lists and chat completions. 
- Avoid reading huge response bodies into memory and fail fast when `Content-Length` declares a response larger than a sensible limit. 
- Ensure error responses are safely reported without spamming terminals or logs with raw control characters or multi-megabyte blobs.

### Description

- Add per-purpose limits: `ModelsResponseMaxBytes = 2 MiB`, `ChatResponseMaxBytes = 4 MiB`, `GeneratedContentMaxCharacters = 1 Mi` and `ErrorBodyMaxBytes = 4 KiB` in `LocalAiProvider`. 
- Use `HttpCompletionOption.ResponseHeadersRead` for `/models` and `/chat/completions`, check `Content-Length` before reading, and reject oversized declared responses. 
- Stream-parse JSON directly via `ParseResponseAsync` which wraps the response stream in a `LimitedReadStream` sentinel to stop parsing once the byte limit is exceeded and avoids building huge intermediate strings. 
- Read at most `ErrorBodyMaxBytes` from non-success responses with `ReadErrorSnippetAsync`, truncate and escape control characters via `TruncateAndEscape`, and cap the final `choices[0].message.content` length with an explicit error. 
- Add unit tests that exercise oversized `Content-Length` rejection, chunked/unknown-length stream overflow detection, large error-body truncation/escaping, an in-limit successful response, and generated-content character-limit rejection in `tests/Wip.Tests/WipAiTests.cs`.

### Testing

- `git diff --check` was run to validate formatting and reported no issues. 
- `dotnet test tests/Wip.Tests/Wip.Tests.csproj --no-restore` was attempted but could not be executed in this environment because `dotnet` is not installed, so the new tests were added but not executed here.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_6a975477d2288322920ea296cf30931d)

## Summary by CodeRabbit

* **New Features**
  * Added time limits for model discovery and chat generation, including streamed responses.
  * Preserved caller cancellation during chat generation.
  * Added safeguards limiting response sizes and generated-content length.
  * Improved handling of malformed responses and safely truncated error details.

* **Bug Fixes**
  * Prevented excessive response data from being read during AI interactions.

* **Tests**
  * Added coverage for response limits, streaming, error reporting, and content-length boundaries.

* **Chores**
  * Updated the project version to 2.3.2.

### CODE REVIEW

File: src/Wip.Core/Ai/LocalAiProvider.cs [hunk 1]
```diff
@@ -97,7 +101,8 @@ public static string DiscoverModel(string baseUrl, HttpMessageHandler? handler =
         HttpResponseMessage response;
         try
         {
-            response = client.GetAsync($"{baseUrl}/models").GetAwaiter().GetResult();
+            response = client.GetAsync(
+                $"{baseUrl}/models", HttpCompletionOption.ResponseHeadersRead).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (abechan1) [reply src/Wip.Core/Ai/LocalAiProvider.cs]: Fixed in dbe27d9.

`DiscoverModel` now creates a `CancellationTokenSource(timeout)` before the request and passes its token to `GetAsync`, `ReadErrorSnippetAsync` and `ParseResponseAsync`, so headers and body share one 5-second budget. `Generate` uses `CancellationTokenSource.CreateLinkedTokenSource(cancellationToken)` with `CancelAfter(5 minutes)`, so the caller's cancellation is preserved and still propagates unwrapped (the `when (!cancellationToken.IsCancellationRequested)` filter now also guards the body-read block, which is wrapped in its own try/catch so a deadline hit surfaces as the usual "did not respond in time" `WipException`).

Regression tests added: `DiscoverModelGivesUpWhenTheBodyNeverFollowsItsHeaders` and `GenerateGivesUpWhenTheBodyNeverFollowsItsHeaders` serve headers and then a stream that never yields a byte, plus `GenerateLetsCallerCancellationOfAStalledBodyPropagateUnwrapped` for the caller-token path. The deadlines are injectable for tests only (an internal `DiscoverModel` overload and an internal `RequestTimeout` init property) so the tests run in milliseconds instead of waiting out the real 5-minute deadline.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/Wip.Core/Ai/LocalAiProvider.cs [hunk 2]
```diff
@@ -128,50 +135,37 @@ public static string DiscoverModel(string baseUrl, HttpMessageHandler? handler =
         }
     }
 
-    private static IReadOnlyList<string> ExtractModelIds(string responseBody)
+    private static IReadOnlyList<string> ExtractModelIds(JsonDocument document)
     {
-        JsonDocument document;
-        try
+        if (!document.RootElement.TryGetProperty("data", out var data))
```
[the remark this answers is not in the document]
Comment (abechan1) [reply src/Wip.Core/Ai/LocalAiProvider.cs:170]: Fixed in dbe27d9.

`ExtractModelIds` rejects a non-object root before touching `data` (or `error`), and `ExtractContent` checks the root's `ValueKind` and `choices[0].ValueKind` before reading nested properties. Malformed array/string/null bodies now raise a `WipException`, so `Doctor.CheckAi` reports its warning instead of crashing. Covered by `DiscoverModelRejectsARootThatIsNotAnObject`, `GenerateRejectsARootThatIsNotAnObject` and `GenerateRejectsAChoiceThatIsNotAnObject`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/Wip.Core/Ai/LocalAiProvider.cs [hunk 3]
```diff
@@ -97,7 +101,8 @@ public static string DiscoverModel(string baseUrl, HttpMessageHandler? handler =
         HttpResponseMessage response;
         try
         {
-            response = client.GetAsync($"{baseUrl}/models").GetAwaiter().GetResult();
+            response = client.GetAsync(
+                $"{baseUrl}/models", HttpCompletionOption.ResponseHeadersRead).GetAwaiter().GetResult();
         }
         catch (HttpRequestException exception)
```
[the remark this answers is not in the document]
Comment (abechan1) [reply src/Wip.Core/Ai/LocalAiProvider.cs:121]: Fixed in dbe27d9 — same change as the CodeRabbit thread on this file.

Both paths now build a `CancellationTokenSource` before the request and use its token for `GetAsync`/`SendAsync`, `ReadErrorSnippetAsync` and `ParseResponseAsync`, so the body reads share the same budget as the headers instead of running with `CancellationToken.None`. `Generate`'s source is linked to the caller's token, so caller cancellation still propagates unwrapped while a deadline hit becomes the usual "did not respond in time" `WipException`. Regression tests cover a response whose headers arrive but whose body never does.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/Wip.Core/Ai/LocalAiProvider.cs [hunk 4]
```diff
@@ -193,61 +222,164 @@ public string Generate(string prompt, CancellationToken cancellationToken = defa
                 new HttpRequestMessage(HttpMethod.Post, $"{baseUrl}/chat/completions")
                 {
                     Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json"),
-                },
-                cancellationToken).GetAwaiter().GetResult();
+                }, HttpCompletionOption.ResponseHeadersRead,
+                deadline.Token).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (abechan1) [reply src/Wip.Core/Ai/LocalAiProvider.cs]: Fixed in 87706bf. The request is now `using var request = new HttpRequestMessage(...)` at method scope and passed to `SendAsync`, so it (and its `StringContent`) is disposed deterministically. Method scope rather than a block around the send on purpose: the response body is read further down, so tearing the request down early would be the riskier shape.

---
_Generated by [Claude Code](https://claude.ai/code)_
